### ISSUE TITLE
incorrect visualization of initial states

### ISSUE THREAD

**Issue by dsb66** [theislab/cellrank#714]
Cellrank incorrectly visualizes the location of cells in the 'initial states' calculation. I have attached an example of the location of two clusters identified as initial states by cellrank. At left is a standard visualization of the individual clusters using the DimPlot function in Seurat. At right is the visualization of the same clusters by cellrank.  There are only 69 damage-associated 1 cells and they are mostly found in a well-defined spot in the standard umap visualization. Cellrank visualizes what appears to be a much larger number of cells as damage-associated 1, and they are dispersed all over the map. It also displays a much smaller number of cycling cells_G2M, though they are in the correct location on the map.

This incorrect visualization of the cells identified as initial states  does not give me confidence that the identification is correct.

When running the code below, cellrank also returns the warning: "The following states could not be mapped uniquely: `['cycling cells_G2M', 'damage-associated 1']`

![example](https://user-images.githubusercontent.com/47340110/129678370-e7fd57ae-cbfe-4095-9914-7ee5fe794b60.png)

```python
adata = adata_d[sample]
cr.tl.initial_states(adata, cluster_key=anno, n_jobs=20,
                     estimator=cr.tl.estimators.CFLARE,
                     n_states = init_states[sample], weight_connectivities = wc)
cr.pl.initial_states(adata, basis=basis, discrete=True,
                     save='_'.join((sample,mode,basis,anno)) + '_cr_initial_states.pdf')
```

#### Versions:

cellrank==1.4.0 scanpy==1.7.2 anndata==0.7.6 numpy==1.21.0 numba==0.52.0 scipy==1.7.0 pandas==1.3.0 pygpcca==1.0.2 scikit-learn==0.24.2 statsmodels==0.12.2 python-igraph==0.9.1 scvelo==0.2.3 pygam==0.8.0 matplotlib==3.4.2 seaborn==0.11.1

**Comment by Marius1311** [comment]
Hi @dsb66, thanks for using CellRank. There is no bug here, this relates to how we define and label initial states. There are two separate things here:
- clusters, defined though whatever clustering algorithm (e.g. louvain)
- initial states, computes in CellRank

Let's look into the latter a bit more. 

## Definition of macrostates
Initial states, in CellRank, are certain macrostates which fulfill some condition with respect to their metastability. Macrostates are the general concept, these include all of initial, terminal and intermediate states. We define macrostates though soft assignments of cells, i.e. think of it like a soft clustering. You can define this though an assignment matrix `M`, where each row corresponds to a cell and each column corresponds to a macrostate. Entries of `M` are restricted to be positive and sum to one row-wise, i.e. you may interpret the entries of `M` as soft assignments of cells to macrostates. 

## Definition of initial states
So let's say that CellRank detects macrostate `m_1` to be an initial one. How do we represent this state? The most natural choice would be to use the corresponding column of matrix `M`. This column contains the soft weights assigning cells to `m_1`. However, let's say we want to represent this initial state through a discrete set of cells, as you did it above. For that, we will simply select the `n` cells from `m_1` that have the highest weight assigned, i.e. these are the cells that we are most confident about that they will belong to `m_`. 

## Naming and coloring and initial state
So far, we managed to find the initial state, and we selected `n` cells that we are pretty confident in assigning to this initial state. Great! But what name are we going to give to this state? Just `m1`? That's not very informative. Instead, we're going to use the independent cluster annotations which are present in your data object. Now, we need to match our `n` cells with any of the predefined clusters. Sometimes, this will work really well and all `n` cells come from the same cluster. In this case, the initial state will visually agree with the cluster whose name it inherited. In other cases, it may well happen that we distribute the `n` cells across several clusters. Now, which name are we going to give to that state? For simplicity, we'll just give it the name corresponding to the cluster we share *most cells* with, i.e. the one we have the largest overlap with. We'll also associate that color with the initial state. CellRank automatically detects cases in which the `n` cells corresponding to your initial state are distributed across several clusters and will issue a warning for this (`The following states could not be mapped uniquely...`).

## Conclusion
There's nothing "wrong" here, everything behaves as expected. The problem is that your initial state is broadly distributed across several clusters, so it's hard to associate it with any particular one for naming. CellRank automatically detected this and gave you a warning. Note that this only relates to the name and the color of that state - everything else is entirely unaffected.

## Side notes
I take from your post that you aren't very happy about the initial states you got. Please consider using our low level interface via kernels and estimators to get more control of the computations we perform under the hood. In particular, I encourage you to go through the  [kernels and estimators tutorial](https://cellrank.readthedocs.io/en/stable/kernels_and_estimators.html). Also, please consider using the GPCCA estimator, it's far more advanced than CFLARE and will likely give you much better results.

**Comment by dsb66** [comment]
Thank you for the explanation, it makes very much sense. I have also started playing with the low level interface as you suggested and I appreciated the more granular control (and understanding) of the process.

I still have a question about the labeling of the states. I produced the attached terminal states map with this code:

    g.compute_macrostates(n_cells=50, cluster_key=anno)
    g.plot_macrostates(discrete=True,
                       save='_'.join((sample,mode,basis,anno)) + '.terminal_states.pdf')

The position of the label for 'enterocytes 2' is correct, but the labels for 'cycling cells S' and 'secretory 2' are swapped: the cells at the location of the 'secretory 2' label are all cycling cells and vice-versa. This also affects the plot of individual state probabilities done by passing same_plot=False.

[scvelo_CT_dynamical_umap_cell.type_2_PA.ex.terminal_states.pdf](https://github.com/theislab/cellrank/files/7003931/scvelo_CT_dynamical_umap_cell.type_2_PA.ex.terminal_states.pdf)

**Comment by Marius1311** [comment]
@michalk8, any idea what could have happened here? I'm not sure this is actually a bug, we need to figure out what's going on.

**Comment by michalk8** [comment]
> @michalk8, any idea what could have happened here? I'm not sure this is actually a bug, we need to figure out what's going on.

@dsb66 can you please post the output of the following?
```python
import scvelo as scv
scv.pl.scatter(g.adata, color=anno)
```
My suspicion is that we for some reason map the names incorrectly (or a problem with the `Lineage` class),
but wanna rule any plotting issues first.

**Comment by dsb66** [comment]
Sorry for the slow reply. I have attached two files, they are from a different dataset but they show the same problem.

The file 'scvelo_CT_dynamical_umap_cell.type_2_PA.ex.macrostates.pdf' was generated by CellRank with the code in my previous post; it shows labels in swapped positions.

The file 'scvelo_scv.pl_CT_dynamical_umap_cell.type_2_PA.ex.png' was generated with the code:

scv.pl.scatter(g.adata, color=anno, save='scv.pl_CT_dynamical_umap_cell.type_2_PA.ex.png')

with all the labels at the correct locations.




![scvelo_scv pl_CT_dynamical_umap_cell type_2_PA ex](https://user-images.githubusercontent.com/47340110/130540365-9c249a27-f411-410f-856b-22cc83f52288.png)
[scvelo_CT_dynamical_umap_cell.type_2_PA.ex.macrostates.pdf](https://github.com/theislab/cellrank/files/7035752/scvelo_CT_dynamical_umap_cell.type_2_PA.ex.macrostates.pdf)

**Comment by MubasherMohammed** [comment]
Hi guys, @dsb66 @michalk8 
i have the same issue in swapping the labels of the terminal states, i haven't manually annotate the clusters they are just inherited from louvain clustering.  
![image](https://user-images.githubusercontent.com/50444056/130768104-c5967a49-f0b8-4fcb-a68c-b0bf7a2c61f8.png)
![image](https://user-images.githubusercontent.com/50444056/130768172-4754b418-4e7d-49ef-a64e-181eb9f94423.png)

have you guys manage to resolve this?
TIA

**Comment by michalk8** [comment]
@MubasherMohammed @dsb66 seems like the wrong name has been assigned, but the color is right (or maybe just the color array in the `adata` is being somehow flipped). Could maybe one of you print the output of the following code?
```python3
# after running g.compute_macrostates(cluster_key=...)
print(list(g.macrostates.cat.categories))
print(list(g._macrostates_colors))
print(list(g.macrostates_memberships.names))
print(list(g.macrostates_memberships.colors))

g.set_terminal_states_from_macrostates()
print(list(g.terminal_states.cat.categories))
print(list(adata.obs['terminal_states'].cat.categories))
print(list(adata.uns['terminal_states_colors']))
```

**Comment by MubasherMohammed** [comment]
Hi @michalk8 thanks for fast reply.
here the output of previous codes 
`print(list(g.macrostates.cat.categories))
print(list(g._macrostates_colors))
print(list(g.macrostates_memberships.names))
print(list(g.macrostates_memberships.colors))
['0', '5']
['#8c564b', '#1f77b4']
['0', '5']
['#8c564b', '#1f77b4']
g.set_terminal_states_from_macrostates()
print(list(g.terminal_states.cat.categories))
print(list(adata.obs['terminal_states'].cat.categories))
print(list(adata.uns['terminal_states_colors']))
['0', '5']
['0', '5']
['#8c564b', '#1f77b4']`

**Comment by Marius1311** [comment]
Hi @MubasherMohammed, if at all possible, could you somehow enable us to reproduce this by sharing an AnnData object and jupyter notebook? I understand if that's not possible due to confidentiality etc. just wondering. Feel free to email me at marius.lange@helmholtz-muenchen.de

**Comment by Marius1311** [comment]
@MubasherMohammed, is your issue also resolved on the dev branch?

**Comment by MubasherMohammed** [comment]
@Marius1311  Yes, it works with the dev branch. Thanks guys

**Comment by Marius1311** [comment]
@michalk8, it looks like we can close this!

**Comment by michalk8** [comment]
> @michalk8, it looks like we can close this!

I don't know why it's been fixed though - can't remember any changes on `dev` that would lead to a fix, will keep this open until I get to investigate.

**Comment by michalk8** [comment]
The only relevant change that might be relevant (although afaik, the results should be identical), is
```python
-        colors_query = []
-        for name in names_query:
-            mask = cats_reference == name
-            color = np.array(colors_reference)[mask][0]
-            colors_query.append(color)
-        association_df["color"] = colors_query
+        association_df["color"] = colors_query = [
+            colors_reference[np.where(cats_reference == name)[0][0]]
+            for name in names_query
+        ]
```
in `cellrank/tl/_colors`, function `_map_names_and_colors`.
In any case, there are at least 3 functions (this, `_merge_categorical_series` and `_fuzzy_to_discrete` that should be refactored and simplified in order to prevent this from happening). @Marius1311 since you're the most familiar with these, you should have the honor.

**Comment by Marius1311** [comment]
ok, please open a separate issue for this and I'll have a look.

**Comment by michalk8** [comment]
If this issue comes-up again, I will reopen this (will be fixed in #745).
closed via https://github.com/theislab/cellrank/commit/c8c2b9f6bd2448861fb414435aee7620ca5a0bad
